Defines code review & approval roles

mrjoro opened the pull request:
This documents the TSC's definition of code review approval/merging roles as described in ampproject/meta-tsc#3.

Since this seemed (a) specific to amphtml and (b) not really about governance I removed the earlier versions of these roles from ampproject/meta/GOVERNANCE.md and added these definitions in the code contribution docs for amphtml.

(Other repos that want to follow the same process can link to this definition.  Repos that want to use a different process can define them in their CONTRIBUTING.md)

/cc @ampproject/tsc @erwinmombay @rsimha @tobie

mrjoro commented:
One thing I'd be curious to hear suggestions on--this document previously referred to "reviewers" as a more general term for people who would help guide someone through the process of getting the feature approved.  To avoid confusion I'm planning on updating this term--perhaps to "guide"?

cramforce commented on contributing/contributing-code.md:
```diff
@@ -122,3 +134,43 @@ We recommend you handle signing/being covered by a CLA *before* you send a PR to
+  * A person may become a Reviewer after 10 merged PRs or 10 high quality reviews of complex PRs and a +1 from 1 current REVIEWER.  Qualifying PRs must be non-trivial (not only fixing typos, not only config changes) and should have implemented or documented at least 2 new features.
+  * The list of Reviewers is maintained in the [Reviewers (amphtml)](https://github.com/orgs/ampproject/teams/reviewers-amphtml) GitHub team.
+
+* Collaborators
```
I'd move this up. At least above reviewers, and maybe above OWNERS

mrjoro replied:
Done; I've switched the order to Collaborators, Reviewers then Owners.

cramforce commented on contributing/contributing-code.md:
```diff
@@ -122,3 +134,43 @@ We recommend you handle signing/being covered by a CLA *before* you send a PR to
+  * Review, approve and merge PRs in the repo for which COLLABORATOR status is granted.
+  * Collaborator status is granted to folks who have proven basic familiarity with the respective repository.
+  * A person may become a Collaborator after 2 merged PRs that are non-trivial (not only fixing typos, not only config changes) and a +1 from 1 current Reviewer.
+  * The list of Collaborators is maintained in the [Collaborators (amphtml)](https://github.com/orgs/ampproject/teams/collaborators-amphtml) GitHub team.
```
We need to document how to apply!

mrjoro replied:
Good point... how do we want to handle that?  File an issue in the given repo (e.g. amphtml) and cc ampproject/collaborators-amphtml ?

cramforce replied:
Sorry, missed this comment. I think filing an issue is good but CCing the full list is probably too much. Probably better to CC reviewers.

mrjoro replied:
I updated each of the roles to document how to request access.  PTAL.

tobie commented on contributing/contributing-code.md:
```diff
@@ -122,3 +134,43 @@ We recommend you handle signing/being covered by a CLA *before* you send a PR to
+* Owners
+  * Review & approve PRs in the area in which they have expertise.
+  * Requirements to be an Owner:
+    * Demonstrated expertise in the area in which they are an OWNER.
```
What's the distinction between "owner", "Owner", and "OWNER"? If there isn't any, pick one for consistency. I'd suggest going for the lowercased version. The less this look like it's been written by a lawyer, the more engineers will read it.

mrjoro replied:
There is no significance to the different forms; I've updated them.

mrjoro commented on contributing/contributing-code.md:
```diff
@@ -122,3 +134,43 @@ We recommend you handle signing/being covered by a CLA *before* you send a PR to
+  * The list of Owners for a directory can be found in the [OWNERS.yaml](https://github.com/ampproject/amphtml/search?o=asc&q=filename%3AOWNERS.yaml&s=indexed) file in the directory or a parent directory.
+
+* Reviewers
+  * Review, approve and merge PRs in the repo for which Reviewer status is granted.
```
I've made this change as part of the other changes.

mrjoro commented on contributing/contributing-code.md:
```diff
@@ -122,3 +134,43 @@ We recommend you handle signing/being covered by a CLA *before* you send a PR to
+* Reviewers
+  * Review, approve and merge PRs in the repo for which Reviewer status is granted.
+  * Reviewer status is granted to folks who have demonstrated deep familiarity with the code-style and conventions of the respective repository.
+  * A person may become a Reviewer after 10 merged PRs or 10 high quality reviews of complex PRs and a +1 from 1 current REVIEWER.  Qualifying PRs must be non-trivial (not only fixing typos, not only config changes) and should have implemented or documented at least 2 new features.
```
Done; thanks!

mrjoro commented on contributing/contributing-code.md:
```diff
@@ -122,3 +134,43 @@ We recommend you handle signing/being covered by a CLA *before* you send a PR to
+
+### Approval requirements
+
+To be merged, a PR must always be reviewed and approved by:
```
Done.

mrjoro commented on contributing/contributing-code.md:
```diff
@@ -122,3 +134,43 @@ We recommend you handle signing/being covered by a CLA *before* you send a PR to
+
+It is acceptable for one person to fulfill these requirements, e.g. if an Owner who is also a Reviewer approves the PR it may be merged.
+
+Any Reviewer or Collaborator may merge the code after it has been approved, including the PR author.
```
Done.

mrjoro commented on contributing/contributing-code.md:
```diff
@@ -122,3 +134,43 @@ We recommend you handle signing/being covered by a CLA *before* you send a PR to
+
+### Roles
+
+* Owners
```
Done.

mrjoro commented:
@tobie regarding your comments:

- I've updated the doc to be more clear about the purpose of the different roles in the first section.  This also made the section on what reviews were required for approval to fit more naturally there.
- I fixed the unintended capitalization issues.
- To a certain extent we're using tooling already with these roles to get to the permissions we want while working with GitHub's permissions model--e.g. the distinction between Collaborator and Reviewer will be enforced by a bot since both have merge access.
- I agree losing the "contributor" role has some drawbacks, but it felt like it might add some confusion since this section is really about "what do I need to do to get my code in."
- Having a single set of Reviewers across repositories is an interesting idea, especially since we do have OWNERS that handles the "but this repo doesn't really have anything to do with amphtml."  I'm curious what @cramforce and the rest of the @ampproject/tsc think.

tobie commented on contributing/contributing-code.md:
```diff
@@ -90,8 +100,10 @@ Once you have found a reviewer, make sure to @-mention them on any issues / PRs
+  - Add reviewers to your PR that will fulfill the requirements of code review and approval documented in the [Code review and approval](#code-review-and-approval) section.  (Your reviewer can help with this.)
   - [Respond to feedback](./getting-started-e2e.md#respond-to-pull-request-comments).
-- After your PR is approved, it will merged by your reviewer or one of the other approvers. To check on your changes and find out when they get into production, read [See your changes in production](./getting-started-quick.md#see-your-changes-in-production).
+- After your PR is approved, it will merged by one of the Reviewers/Collaborators who reviewed your code.
```
Couldn't we use the same language here as below, e.g.: "Once the PR has been approved, anyone with commit rights to the repository may merge the PR, including its author if they have commit rights."

mrjoro replied:
Since this is from the perspective of someone going through the review process I wanted to make it a bit more actionable--i.e. if the code is not merged what should they do?  I've updated it to make this more obvious.

tobie commented:
@mrjoro thanks for addressing my comments and apologies for their overall tone. I just re-read the conversation right now and I really don't come across as I intended to. Sorry. :-/

> Having a single set of Reviewers across repositories is an interesting idea, especially since we do have OWNERS that handles the "but this repo doesn't really have anything to do with amphtml." I'm curious what @cramforce and the rest of the @ampproject/tsc think.

That's not _really_ what I was suggesting, btw, but it's an interesting idea nonetheless.

My suggestion was to keep the _definition_ of common roles in the governance doc, and have projects define the _path_ to these roles (i.e. what you need to have done to prove yourself worthy) for that particular repo in the contributing doc of the repo itself.

dvoytenko commented on contributing/contributing-code.md:
```diff
@@ -90,8 +100,10 @@ Once you have found a reviewer, make sure to @-mention them on any issues / PRs
   - [Push your changes](./getting-started-e2e.md#push-your-changes-to-your-github-fork)
   - [Create a Pull Request (PR)](./getting-started-e2e.md#send-a-pull-request-ie-request-a-code-review).
   - Make sure the presubmit checks shown on your PR on GitHub passes (e.g. no lint and type check errors, tests are passing).
+  - Add reviewers to your PR that will fulfill the requirements of code review and approval documented in the [Code review and approval](#code-review-and-approval) section.  (Your reviewer can help with this.)
```
Is it now possible to add reviewers to PRs for folk w/o write access?

erwinmombay replied:
i believe it is not, as far as i know

mrjoro replied:
Yes, unfortunately they can't.  That's why I added "your reviewer can help with this."

erwinmombay commented on contributing/contributing-code.md:
```diff
@@ -122,3 +134,41 @@ We recommend you handle signing/being covered by a CLA *before* you send a PR to
+To be merged, all code must be approved by both:
+
+* At least one [Reviewer](https://github.com/orgs/ampproject/teams/reviewers-amphtml) who is not the author.  If the author is a Reviewer, a [Collaborator](https://github.com/orgs/ampproject/teams/reviewers-amphtml) may fulfill this requirement instead.
+* At least one [Owner](https://github.com/ampproject/amphtml/search?o=asc&q=filename%3AOWNERS.yaml&s=indexed) for all areas the PR affects, except those areas in which the code author is an Owner.
```
currently if the author is also an owner, this will also fulfill owners approval. it will still need a `github` reviewer approval (seperate from the reviewer mentioned above)

mrjoro replied:
Right, that's what I was trying to get across with "except those areas in which the code author is an Owner."  Should I make this more clear?

erwinmombay replied:
after re-reading it is pretty clear. wording just isnt consistent with the `reviewer` section above which threw me off a bit

mrjoro commented:
@tobie regarding the definition of roles vs. path to getting to that role, I'd prefer to keep them together for now to avoid someone having to look in multiple places to get an understanding of how things work.  We point new contributors to this contributing-code.md document for how to get their code in so it feels like a natural place for it, especially since by far most contributors are submitting code to amphtml.

If we find that different repos are using the same roles concepts but different ways of getting those roles then we can revisit this.

WDYT?

mrjoro commented:
Friendly ping @tobie @dvoytenko @erwinmombay @cramforce :)

cramforce commented on contributing/contributing-code.md:
```diff
@@ -36,10 +36,20 @@ Significant changes (e.g. new components or significant changes to behavior) req
+If you're making a minor fix and just want to find someone to review/approve your code, choose:
+
+- an Owner from the OWNERS.yaml file in the directories you change (or their parent directories)
+- and a [Reviewer](https://github.com/orgs/ampproject/teams/reviewers-amphtml)
```
One "bug" in our approach here is that these teams are not, and cannot be made publicly visible.

I wonder if we should spend the time to write a quick web app that publishes the teams on the web?

mrjoro replied:
I filed https://github.com/ampproject/meta/issues/32 to track adding this.  I figure we can get this PR merged and handle that separately.

erwinmombay replied:
Since the owners bot already fetches the teams information on each pull request this information can easily be exposed in the checks tab but i dont know if that is the correct place to enumerate them.

cramforce replied:
Joey and I discussed that the best place is really to just put it on amp.dev which should be super straight forward. Part of the site is already generated from GH API calls.

mrjoro commented:
@tobie I'll go ahead and submit this.  We can continue discussing whether it makes sense to move this back up to a higher level separately and handle that in a future PR if we decide to change it.

tobie commented:
Sorry for not getting back to you earlier @mrjojo. I just got back from London this evening and was hoping to comment on this early next week. Your approach sounds good, though; let’s revisit if/when we need to.
